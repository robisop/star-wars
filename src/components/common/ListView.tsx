import {ListItemData} from 'api/common';
import {ListItem} from 'components/common/ListItem';
import {StatusBar} from 'expo-status-bar';
import {
  ActivityIndicator,
  FlatList,
  StyleSheet,
  Text,
  View,
} from 'react-native';

interface ListViewProps {
  data: ListItemData[] | undefined;
  hasNextPage: boolean | undefined;
  isLoading: boolean;
  isError: boolean;
  isFetchingNextPage: boolean;
  fetchNextPage: () => void;
}

export const ListView = ({
  data,
  hasNextPage,
  isLoading,
  isError,
  isFetchingNextPage,
  fetchNextPage,
}: ListViewProps) => {
  if (isLoading) {
    return (
      <View style={styles.container}>
        <Text>Loading...</Text>
        <StatusBar style="auto" />
      </View>
    );
  }

  if (isError) {
    return (
      <View style={styles.container}>
        <Text>An error has occurred</Text>
        <StatusBar style="auto" />
      </View>
    );
  }

  const renderFooter = () => (
    <View style={styles.listFooter}>
      {hasNextPage && isFetchingNextPage && (
        <ActivityIndicator testID="activityIndicator" />
      )}
    </View>
  );

  const renderEmpty = () => (
    <View style={styles.listEmpty}>
      <Text>No data</Text>
    </View>
  );

  return (
    <View style={styles.container}>
      <FlatList
        testID="flatList"
        data={data}
        renderItem={({item}) => (
          <ListItem title={item.title} subtitle={item.subtitle} />
        )}
        keyExtractor={item => item.id}
        ListFooterComponent={renderFooter}
        ListEmptyComponent={renderEmpty}
        onEndReachedThreshold={0.5}
        onEndReached={() => (hasNextPage ? fetchNextPage() : undefined)}
      />
      <StatusBar style="auto" />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
  listFooter: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  listEmpty: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
