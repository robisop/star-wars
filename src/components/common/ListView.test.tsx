import {render, screen, fireEvent} from '@testing-library/react-native';

import {ListView} from './ListView';

const baseProps = {
  data: [
    {
      id: '1',
      title: 'test title',
      subtitle: 'test subtitle',
    },
    {
      id: '2',
      title: 'test title 2',
      subtitle: 'test subtitle 2',
    },
  ],
  hasNextPage: false,
  isLoading: false,
  isError: false,
  isFetchingNextPage: false,
  fetchNextPage: jest.fn(),
};

const scrollEventData = {
  nativeEvent: {
    contentOffset: {
      y: 500,
    },
    contentSize: {
      // Dimensions of the scrollable content
      height: 500,
      width: 100,
    },
    layoutMeasurement: {
      // Dimensions of the device
      height: 100,
      width: 100,
    },
  },
};

describe('<ListView />', () => {
  it('renders empty list correctly', () => {
    const props = {
      ...baseProps,
      data: [],
    };

    render(<ListView {...props} />);
    expect(screen.getByText('No data')).toBeDefined();
  });

  it('renders loading state correctly', () => {
    const props = {
      ...baseProps,
      isLoading: true,
    };

    render(<ListView {...props} />);
    expect(screen.getByText('Loading...')).toBeDefined();
  });

  it('renders error state correctly', () => {
    const props = {
      ...baseProps,
      isError: true,
    };

    render(<ListView {...props} />);
    expect(screen.getByText('An error has occurred')).toBeDefined();
  });

  it('renders data correctly', () => {
    render(<ListView {...baseProps} />);
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('renders fetching state correctly', () => {
    const props = {
      ...baseProps,
      hasNextPage: true,
      isFetchingNextPage: true,
    };

    render(<ListView {...props} />);
    expect(screen.getByTestId('activityIndicator')).toBeDefined();
  });

  it('calls fetchNextPage when scrolled to the end of list and hasNextPage is true', () => {
    const props = {
      ...baseProps,
      hasNextPage: true,
      fetchNextPage: jest.fn(),
    };

    render(<ListView {...props} />);
    fireEvent.scroll(screen.getByTestId('flatList'), scrollEventData);
    expect(props.fetchNextPage).toHaveBeenCalled();
  });

  it('does not call fetchNextPage when scrolled to the end of list and hasNextPage is false', () => {
    const props = {
      ...baseProps,
      hasNextPage: false,
      fetchNextPage: jest.fn(),
    };

    render(<ListView {...props} />);
    fireEvent.scroll(screen.getByTestId('flatList'), scrollEventData);
    expect(props.fetchNextPage).not.toHaveBeenCalled();
  });
});
