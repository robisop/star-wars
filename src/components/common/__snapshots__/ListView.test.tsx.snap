// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListView /> renders data correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    data={
      [
        {
          "id": "1",
          "subtitle": "test subtitle",
          "title": "test title",
        },
        {
          "id": "2",
          "subtitle": "test subtitle 2",
          "title": "test title 2",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    testID="flatList"
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "marginBottom": 10,
              "minWidth": "100%",
              "paddingBottom": 10,
              "paddingLeft": 30,
              "paddingRight": 10,
              "paddingTop": 20,
            }
          }
        >
          <View
            style={
              {
                "marginBottom": 5,
                "marginTop": 5,
              }
            }
          >
            <Text
              style={
                {
                  "fontWeight": "bold",
                }
              }
            >
              test title
            </Text>
          </View>
          <View
            style={
              {
                "marginBottom": 10,
              }
            }
          >
            <Text>
              test subtitle
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "marginBottom": 10,
              "minWidth": "100%",
              "paddingBottom": 10,
              "paddingLeft": 30,
              "paddingRight": 10,
              "paddingTop": 20,
            }
          }
        >
          <View
            style={
              {
                "marginBottom": 5,
                "marginTop": 5,
              }
            }
          >
            <Text
              style={
                {
                  "fontWeight": "bold",
                }
              }
            >
              test title 2
            </Text>
          </View>
          <View
            style={
              {
                "marginBottom": 10,
              }
            }
          >
            <Text>
              test subtitle 2
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
